prow/plugins/projectmanager/projectmanager.go

```diff
@@ -0,0 +1,204 @@
+				}
+				configString = fmt.Sprintf("%s }", configString)
+			}
+			configString = fmt.Sprintf("%s }", configString)
+			configMap[""] = configString
+			return configMap
+		}(config),
+	}
+	return pluginHelp, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pe github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.PluginConfig.ProjectManager, pc.Logger, pe)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetRepoProjects(owner, repo string) ([]github.Project, error)
+	GetOrgProjects(org string) ([]github.Project, error)
+	GetProjectColumns(projectID int) ([]github.ProjectColumn, error)
+	CreateProjectCard(columnID int, projectCard github.ProjectCard) (*github.ProjectCard, error)
+}
+
+func handlePR(gc githubClient, projectManager plugins.ProjectManager, log *logrus.Entry, pe github.PullRequestEvent) error {
+	// Only respond to label add or issue/PR open events
+	if pe.Action != github.PullRequestActionOpened &&
+		pe.Action != github.PullRequestActionReopened &&
+		pe.Action != github.PullRequestActionLabeled {
+		return nil
+	}
+	// Get any ManagedProjects that match this PR
+	matchedColumnIDs, err := getMatchingColumnIDs(gc, projectManager, pe)
+	if err != nil {
+		return err
+	}
+	// For each ManagedColumn that matches this PR, add this PR to that Project Column
+	for _, matchedColumnID := range matchedColumnIDs {
+		err = addPRToColumn(gc, matchedColumnID, pe)
+		log.WithError(err).Println("Failed to add PR to project")
+	}
+	return nil
+}
+
+func getMatchingColumnIDs(gc githubClient, projectManager plugins.ProjectManager, pe github.PullRequestEvent) ([]int, error) {
+	var matchedColumnIDs []int
+	// Don't use GetIssueLabels unless it's required and keep track of whether the labels have been fetched to avoid unnecessary API usage.
+	labelsFetched := false
+	var labels []github.Label
+	for orgRepoName, managedOrgRepo := range projectManager.OrgRepos {
+		for projectName, managedProject := range managedOrgRepo.Projects {
+			for _, managedColumn := range managedProject.Columns {
+				if managedColumn.Org != "" && managedColumn.Org != pe.Repo.Owner.Login {
+					continue
+				}
+				if managedColumn.State != "" && managedColumn.State != pe.PullRequest.State {
+					continue
+				}
+				if len(managedColumn.Labels) != 0 {
+					if !labelsFetched {
+						// If labels are not yet fetched then get them as they are now required
+						// GetIssueLabels works for PRs as they are considered issues in the API
+						var err error
+						labels, err = gc.GetIssueLabels(pe.Repo.Owner.Login, pe.Repo.Name, pe.Number)
+						if err != nil {
+							return nil, err
+						}
+						labelsFetched = true
+					}
+					if !hasLabels(managedColumn.Labels, labels) {
+						continue
+					}
+				}
+				columnID := managedColumn.ID
+				// Currently this assumes columnID having a value if 0 means it is unset
```

Comment:
You can use a `*int` here and check `nil` instead of special-casing `0` if you'd like

Reply:
I need to have a default value to look for. It is either 0 or -1 which can never be a valid column id, *int will make me have configuration omitting the column, unless there is a good need for all that code change?

Reply:
Why do you need to have a default value? You are saying you need to check to see if the value is set or not. Using a pointer and checking if it is `nil` is the idiomatic way to check if a value has or has not been set.

Reply:
changed code to use int*